File: reference/provenance-go/receipt.go
package provenance

import (
	"bytes"
	"crypto/ed25519"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"reflect"
	"sort"
)

// This implementation mirrors the canonical Python reference
// (raucle/provenance.py) byte-for-byte: same JOSE header (incl.
// the "raucle/v1": "provenance" tag), same payload field set/ordering,
// same string-typed model/tool/corpus, sha256:-prefixed hashes, and the
// same content-addressed id ("sha256:" + hex(sha256(jws))).

const (
	iss    = "raucle-detect/provenance"
	typ    = "provenance-receipt/v1"
	jwsTyp = typ
)

var validOperations = map[string]bool{
	"user_input": true, "model_call": true, "tool_call": true,
	"retrieval": true, "guardrail_scan": true, "agent_handoff": true,
	"sanitisation": true, "merge": true,
}

var jwsCrit = []string{"raucle/v1"}

// knownFields is the closed set of payload keys a verifier accepts
// (plus x_-prefixed extensions). Mirrors the spec §4 field list.
var knownFields = map[string]bool{
	"iss": true, "typ": true, "iat": true, "agent_id": true,
	"agent_key_id": true, "operation": true, "parents": true,
	"input_hash": true, "output_hash": true, "taint": true,
	"ruleset_hash": true, "guardrail_verdict": true, "model": true,
	"tool": true, "corpus": true, "tenant": true,
}

// Payload is the v1 receipt payload (§4). String-typed hash/model/tool
// fields mirror the canonical Python reference exactly.
type Payload struct {
	Iat              int64
	AgentID          string
	AgentKeyID       string
	Operation        string
	Parents          []string
	Taint            []string
	InputHash        string // empty means absent; value is "sha256:<hex>"
	OutputHash       string // empty means absent
	Model            string // empty means absent
	Tool             string // empty means absent
	Corpus           string // empty means absent
	RulesetHash      string // empty means absent
	GuardrailVerdict string // empty means absent
	Tenant           string // empty means absent; "" tenant set via TenantSet
	TenantSet        bool   // mirrors Python's `tenant is not None`
}

// Receipt is a signed payload.
type Receipt struct {
	JWS     string
	Payload Payload
	// ID is the content-addressed identifier (§8): "sha256:" + hex
	// SHA-256 of the Compact JWS ASCII bytes.
	ID string
}

func b64u(b []byte) string {
	return base64.RawURLEncoding.EncodeToString(b)
}

func b64uDecode(s string) ([]byte, error) {
	return base64.RawURLEncoding.DecodeString(s)
}

func sha256Hex(b []byte) string {
	d := sha256.Sum256(b)
	return hex.EncodeToString(d[:])
}

// Validate enforces the §4 payload rules. It is intentionally lenient in
// the same places the Python reference is: it does not re-shape hashes
// or models, it only enforces structural invariants the spec mandates.
func (p *Payload) Validate() error {
	if !validOperations[p.Operation] {
		return fmt.Errorf("unknown operation: %s", p.Operation)
	}
	if p.Operation == "guardrail_scan" && p.GuardrailVerdict == "" {
		return fmt.Errorf("guardrail_scan requires a verdict (§4)")
	}
	if (p.Operation == "guardrail_scan") && p.RulesetHash == "" {
		return fmt.Errorf("guardrail_scan requires ruleset_hash (§4)")
	}
	if p.Operation == "model_call" && p.Model == "" {
		return fmt.Errorf("model_call requires model (§4)")
	}
	if (p.Operation == "tool_call" || p.Operation == "sanitisation") && p.Tool == "" {
		return fmt.Errorf("%s requires tool (§4)", p.Operation)
	}
	if (p.Operation == "retrieval" || p.Operation == "sanitisation") && p.Corpus == "" {
		return fmt.Errorf("%s requires corpus (§4)", p.Operation)
	}
	if p.Operation == "user_input" && len(p.Parents) > 0 {
		return fmt.Errorf("user_input must have no parents")
	}
	if p.Operation != "user_input" && len(p.Parents) == 0 {
		return fmt.Errorf("%s requires at least one parent", p.Operation)
	}
	// §4.2/§4.3.1: parents and taint MUST be sorted in UTF-16 code-unit order
	// and unique. A strictly-increasing check enforces both at once (a non-
	// conformant emitter could otherwise sign an unsorted/duplicated array).
	for _, f := range []struct {
		name string
		vals []string
	}{{"parents", p.Parents}, {"taint", p.Taint}} {
		for i := 1; i < len(f.vals); i++ {
			if !lessUTF16(f.vals[i-1], f.vals[i]) {
				return fmt.Errorf("%s must be sorted in UTF-16 code-unit order and unique (§4.3.1)", f.name)
			}
		}
	}
	return nil
}

// toMap builds the canonical payload object exactly as Python's
// ProvenanceReceipt.payload() does. Parents and taint are sorted.
func (p *Payload) toMap() map[string]any {
	parents := append([]string(nil), p.Parents...)
	sort.Slice(parents, func(i, j int) bool { return lessUTF16(parents[i], parents[j]) })
	taint := append([]string(nil), p.Taint...)
	sort.Slice(taint, func(i, j int) bool { return lessUTF16(taint[i], taint[j]) })

	parentsAny := make([]any, len(parents))
	for i, v := range parents {
		parentsAny[i] = v
	}
	taintAny := make([]any, len(taint))
	for i, v := range taint {
		taintAny[i] = v
	}

	m := map[string]any{
		"iss":          iss,
		"typ":          typ,
		"iat":          p.Iat,
		"agent_id":     p.AgentID,
		"agent_key_id": p.AgentKeyID,
		"operation":    p.Operation,
		"parents":      parentsAny,
		"taint":        taintAny,
	}
	if p.InputHash != "" {
		m["input_hash"] = p.InputHash
	}
	if p.OutputHash != "" {
		m["output_hash"] = p.OutputHash
	}
	if p.Model != "" {
		m["model"] = p.Model
	}
	if p.Tool != "" {
		m["tool"] = p.Tool
	}
	if p.Corpus != "" {
		m["corpus"] = p.Corpus
	}
	if p.RulesetHash != "" {
		m["ruleset_hash"] = p.RulesetHash
	}
	if p.GuardrailVerdict != "" {
		m["guardrail_verdict"] = p.GuardrailVerdict
	}
	if p.TenantSet {
		m["tenant"] = p.Tenant
	}
	return m
}

func payloadFromMap(m map[string]any) (Payload, error) {
	for k := range m {
		if !knownFields[k] && (len(k) < 2 || k[:2] != "x_") {
			return Payload{}, fmt.Errorf("reserved unknown field: %s", k)
		}
	}
	// typ must equal the spec literal (§4).
	if t, _ := m["typ"].(string); t != typ {
		return Payload{}, fmt.Errorf("payload typ must be %q, got %q", typ, t)
	}
	p := Payload{}
	getStr := func(k string) string {
		if v, ok := m[k].(string); ok {
			return v
		}
		return ""
	}
	if iat, ok := m["iat"].(float64); ok {
		p.Iat = int64(iat)
	}
	p.AgentID = getStr("agent_id")
	p.AgentKeyID = getStr("agent_key_id")
	p.Operation = getStr("operation")
	p.InputHash = getStr("input_hash")
	p.OutputHash = getStr("output_hash")
	p.Model = getStr("model")
	p.Tool = getStr("tool")
	p.Corpus = getStr("corpus")
	p.RulesetHash = getStr("ruleset_hash")
	p.GuardrailVerdict = getStr("guardrail_verdict")
	if tv, ok := m["tenant"]; ok {
		p.TenantSet = true
		if s, ok := tv.(string); ok {
			p.Tenant = s
		}
	}
	if arr, ok := m["parents"].([]any); ok {
		for _, e := range arr {
			if s, ok := e.(string); ok {
				p.Parents = append(p.Parents, s)
			}
		}
	}
	if arr, ok := m["taint"].([]any); ok {
		for _, e := range arr {
			if s, ok := e.(string); ok {
				p.Taint = append(p.Taint, s)
			}
		}
	}
	return p, nil
}

// PayloadFromHarness builds a Payload from a decoded JSON object as used
// by the cross-language conformance harness. It tolerates payloads that
// already carry the constant iss/typ (they are re-injected on emit).
func PayloadFromHarness(m map[string]any) Payload {
	p, _ := payloadFromMap(m)
	return p
}

// Emit signs a payload and returns a Receipt.
func Emit(p Payload, priv ed25519.PrivateKey) (Receipt, error) {
	if err := p.Validate(); err != nil {
		return Receipt{}, err
	}
	header := map[string]any{
		"alg":       "EdDSA",
		"typ":       jwsTyp,
		"kid":       p.AgentKeyID,
		"crit":      toAnySlice(jwsCrit),
		"raucle/v1": "provenance",
	}
	headerB, err := canonicalEncode(header)
	if err != nil {
		return Receipt{}, err
	}
	payloadB, err := canonicalEncode(p.toMap())
	if err != nil {
		return Receipt{}, err
	}
	signingInput := b64u(headerB) + "." + b64u(payloadB)
	sig := ed25519.Sign(priv, []byte(signingInput))
	jws := signingInput + "." + b64u(sig)
	return Receipt{JWS: jws, Payload: p, ID: "sha256:" + sha256Hex([]byte(jws))}, nil
}

// Verify checks a Compact JWS against pub and parses it.
func Verify(jws string, pub ed25519.PublicKey) (Receipt, error) {
	parts := splitN(jws, '.')
	if len(parts) != 3 {
		return Receipt{}, fmt.Errorf("JWS must have three segments")
	}
	headerB, payloadB, sigB := parts[0], parts[1], parts[2]

	hb, err := b64uDecode(headerB)
	if err != nil {
		return Receipt{}, err
	}
	var header map[string]any
	if err := json.Unmarshal(hb, &header); err != nil {
		return Receipt{}, err
	}
	if header["alg"] != "EdDSA" {
		return Receipt{}, fmt.Errorf("unsupported alg: %v", header["alg"])
	}
	if header["typ"] != jwsTyp {
		return Receipt{}, fmt.Errorf("unexpected typ: %v", header["typ"])
	}
	crit, ok := header["crit"].([]any)
	if !ok || len(crit) != 1 || crit[0] != "raucle/v1" {
		return Receipt{}, fmt.Errorf("crit must be exactly ['raucle/v1']")
	}
	if header["raucle/v1"] != "provenance" {
		return Receipt{}, fmt.Errorf("header 'raucle/v1' must be 'provenance'")
	}
	allowedHeaderKeys := map[string]bool{"alg": true, "typ": true, "kid": true, "crit": true, "raucle/v1": true}
	for k := range header {
		if !allowedHeaderKeys[k] {
			return Receipt{}, fmt.Errorf("unexpected JOSE header key: %s", k)
		}
	}

	// Canonical byte-equality (spec v1 §4.3, matches the Python reference): the
	// signature binds the on-wire bytes, but without re-encoding and comparing,
	// a non-canonical header (unsorted keys / extra whitespace) would still
	// verify, admitting byte-different receipts for the same logical content.
	if canonHeader, cerr := canonicalEncode(header); cerr != nil {
		return Receipt{}, cerr
	} else if !bytes.Equal(canonHeader, hb) {
		return Receipt{}, fmt.Errorf("JOSE header is not canonical JSON (JCS)")
	}

	sig, err := b64uDecode(sigB)
	if err != nil {
		return Receipt{}, err
	}
	if !ed25519.Verify(pub, []byte(headerB+"."+payloadB), sig) {
		return Receipt{}, fmt.Errorf("signature invalid")
	}

	pb, err := b64uDecode(payloadB)
	if err != nil {
		return Receipt{}, err
	}
	var pm map[string]any
	if err := json.Unmarshal(pb, &pm); err != nil {
		return Receipt{}, err
	}
	if canonPayload, cerr := canonicalEncode(pm); cerr != nil {
		return Receipt{}, cerr
	} else if !bytes.Equal(canonPayload, pb) {
		return Receipt{}, fmt.Errorf("JWS payload is not canonical JSON (JCS)")
	}
	p, err := payloadFromMap(pm)
	if err != nil {
		return Receipt{}, err
	}
	if err := p.Validate(); err != nil {
		return Receipt{}, err
	}
	if header["kid"] != p.AgentKeyID {
		return Receipt{}, fmt.Errorf("header.kid != payload.agent_key_id (§3)")
	}
	return Receipt{JWS: jws, Payload: p, ID: "sha256:" + sha256Hex([]byte(jws))}, nil
}

// ── small helpers ─────────────────────────────────────────────────

func toAnySlice(ss []string) []any {
	out := make([]any, len(ss))
	for i, s := range ss {
		out[i] = s
	}
	return out
}

func containsAny(xs []any, want string) bool {
	for _, x := range xs {
		if s, ok := x.(string); ok && s == want {
			return true
		}
	}
	return false
}

func splitN(s string, sep byte) []string {
	var out []string
	start := 0
	for i := 0; i < len(s); i++ {
		if s[i] == sep {
			out = append(out, s[start:i])
			start = i + 1
		}
	}
	out = append(out, s[start:])
	return out
}

// equalStringSets is used by the chain verifier.
func equalStringSlices(a, b []string) bool {
	return reflect.DeepEqual(a, b)
}

File: reference/provenance-rs/src/receipt.rs
//! Receipt payload, envelope, emit/verify — §3, §4, §8.
//!
//! Mirrors the canonical Python reference (raucle_detect/provenance.py)
//! byte-for-byte: same JOSE header (incl. the `"raucle/v1": "provenance"`
//! tag), same payload field set, string-typed model/tool/corpus,
//! sha256:-prefixed hashes, and the same content-addressed id
//! (`"sha256:" + hex(sha256(jws))`).

use crate::canonical::canonical_encode;
use base64::Engine;
use ed25519_dalek::{Signature, Signer, SigningKey, Verifier, VerifyingKey};
use serde_json::{json, Map, Value};
use sha2::{Digest, Sha256};

const JWS_TYP: &str = "provenance-receipt/v1";
const ISS: &str = "raucle-detect/provenance";

#[derive(Debug)]
pub struct ProvError(pub String);
impl std::fmt::Display for ProvError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}
impl std::error::Error for ProvError {}

fn err<T>(s: impl Into<String>) -> Result<T, ProvError> {
    Err(ProvError(s.into()))
}

fn known_field(k: &str) -> bool {
    matches!(
        k,
        "iss" | "typ"
            | "iat"
            | "agent_id"
            | "agent_key_id"
            | "operation"
            | "parents"
            | "input_hash"
            | "output_hash"
            | "taint"
            | "ruleset_hash"
            | "guardrail_verdict"
            | "model"
            | "tool"
            | "corpus"
            | "tenant"
    )
}

fn valid_operation(op: &str) -> bool {
    matches!(
        op,
        "user_input"
            | "model_call"
            | "tool_call"
            | "retrieval"
            | "guardrail_scan"
            | "agent_handoff"
            | "sanitisation"
            | "merge"
    )
}

/// A signed receipt.
#[derive(Debug, Clone)]
pub struct Receipt {
    pub jws: String,
    /// The parsed payload object.
    pub payload: Value,
    /// Content-addressed id (§8): "sha256:" + hex SHA-256 of the JWS.
    pub id: String,
}

fn b64u(b: &[u8]) -> String {
    base64::engine::general_purpose::URL_SAFE_NO_PAD.encode(b)
}

fn b64u_decode(s: &str) -> Result<Vec<u8>, ProvError> {
    base64::engine::general_purpose::URL_SAFE_NO_PAD
        .decode(s)
        .map_err(|e| ProvError(format!("base64: {e}")))
}

fn sha256_hex(b: &[u8]) -> String {
    let mut h = Sha256::new();
    h.update(b);
    hex_encode(&h.finalize())
}

fn hex_encode(b: &[u8]) -> String {
    let mut s = String::with_capacity(b.len() * 2);
    for byte in b {
        s.push_str(&format!("{:02x}", byte));
    }
    s
}

fn str_field<'a>(obj: &'a Map<String, Value>, k: &str) -> Option<&'a str> {
    obj.get(k).and_then(|v| v.as_str())
}

/// Validate a payload object against §4. Lenient where the Python
/// reference is lenient: enforces structural invariants (required
/// fields per operation, parent rules, typ literal), not value shapes.
pub fn validate_payload(p: &Value) -> Result<(), ProvError> {
    let obj = match p.as_object() {
        Some(o) => o,
        None => return err("payload must be an object"),
    };

    for k in obj.keys() {
        if !known_field(k) && !k.starts_with("x_") {
            return err(format!("reserved/unknown field: {k} (§14)"));
        }
    }

    if str_field(obj, "typ") != Some(JWS_TYP) {
        return err(format!("payload typ must be {JWS_TYP:?}"));
    }

    let op = str_field(obj, "operation").unwrap_or("");
    if !valid_operation(op) {
        return err(format!("unknown operation: {op}"));
    }

    let has = |k: &str| str_field(obj, k).map(|s| !s.is_empty()).unwrap_or(false);

    if op == "guardrail_scan" && !has("guardrail_verdict") {
        return err("guardrail_scan requires guardrail_verdict (§4)");
    }
    if op == "guardrail_scan" && !has("ruleset_hash") {
        return err("guardrail_scan requires ruleset_hash (§4)");
    }
    if op == "model_call" && !has("model") {
        return err("model_call requires model (§4)");
    }
    if (op == "tool_call" || op == "sanitisation") && !has("tool") {
        return err(format!("{op} requires tool (§4)"));
    }
    if (op == "retrieval" || op == "sanitisation") && !has("corpus") {
        return err(format!("{op} requires corpus (§4)"));
    }

    let parent_count = obj
        .get("parents")
        .and_then(|v| v.as_array())
        .map(|a| a.len())
        .unwrap_or(0);
    if op == "user_input" && parent_count > 0 {
        return err("user_input must have no parents");
    }
    if op != "user_input" && parent_count == 0 {
        return err(format!("{op} requires at least one parent"));
    }

    Ok(())
}

/// Build the canonical payload object from caller-supplied fields,
/// injecting the constant `iss`/`typ` and sorting parents+taint, exactly
/// as Python's `ProvenanceReceipt.payload()` does.
fn normalise_payload(payload: &Value) -> Result<Value, ProvError> {
    let src = payload
        .as_object()
        .ok_or_else(|| ProvError("payload must be an object".into()))?;
    let mut obj: Map<String, Value> = src.clone();
    obj.insert("iss".into(), json!(ISS));
    obj.insert("typ".into(), json!(JWS_TYP));

    // Sort parents + taint to match the canonical reference.
    for key in ["parents", "taint"] {
        if let Some(arr) = obj.get(key).and_then(|v| v.as_array()) {
            let mut v: Vec<String> = arr
                .iter()
                .filter_map(|e| e.as_str().map(|s| s.to_string()))
                .collect();
            v.sort();
            obj.insert(key.into(), json!(v));
        }
    }
    Ok(Value::Object(obj))
}

/// Emit (sign) a payload object.
pub fn emit(payload: &Value, signing_key: &SigningKey) -> Result<Receipt, ProvError> {
    let payload = normalise_payload(payload)?;
    validate_payload(&payload)?;

    let kid = payload
        .get("agent_key_id")
        .and_then(|v| v.as_str())
        .ok_or_else(|| ProvError("missing agent_key_id".into()))?;

    let header = json!({
        "alg": "EdDSA",
        "typ": JWS_TYP,
        "kid": kid,
        "crit": ["raucle/v1"],
        "raucle/v1": "provenance",
    });

    let header_b = canonical_encode(&header).map_err(|e| ProvError(e.0))?;
    let payload_b = canonical_encode(&payload).map_err(|e| ProvError(e.0))?;
    let signing_input = format!("{}.{}", b64u(&header_b), b64u(&payload_b));
    let sig: Signature = signing_key.sign(signing_input.as_bytes());
    let jws = format!("{}.{}", signing_input, b64u(&sig.to_bytes()));
    let id = format!("sha256:{}", sha256_hex(jws.as_bytes()));
    Ok(Receipt { jws, payload, id })
}

/// Verify a Compact JWS against a public key and parse it.
pub fn verify(jws: &str, verifying_key: &VerifyingKey) -> Result<Receipt, ProvError> {
    let parts: Vec<&str> = jws.split('.').collect();
    if parts.len() != 3 {
        return err("JWS must have three segments");
    }
    let (header_b, payload_b, sig_b) = (parts[0], parts[1], parts[2]);

    let header: Value = serde_json::from_slice(&b64u_decode(header_b)?)
        .map_err(|e| ProvError(format!("header parse: {e}")))?;
    if header.get("alg").and_then(|v| v.as_str()) != Some("EdDSA") {
        return err("unsupported alg");
    }
    if header.get("typ").and_then(|v| v.as_str()) != Some(JWS_TYP) {
        return err("unexpected typ");
    }
    let crit_ok = header
        .get("crit")
        .and_then(|v| v.as_array())
        .map(|a| a.len() == 1 && a[0].as_str() == Some("raucle/v1"))
        .unwrap_or(false);
    if !crit_ok {
        return err("crit must be exactly ['raucle/v1']");
    }
    if header.get("raucle/v1").and_then(|v| v.as_str()) != Some("provenance") {
        return err("header 'raucle/v1' must be 'provenance'");
    }
    if let Some(obj) = header.as_object() {
        for k in obj.keys() {
            if !matches!(k.as_str(), "alg" | "typ" | "kid" | "crit" | "raucle/v1") {
                return err("unexpected JOSE header key");
            }
        }
    }

    let sig_bytes = b64u_decode(sig_b)?;
    let sig = Signature::from_slice(&sig_bytes).map_err(|e| ProvError(format!("sig: {e}")))?;
    let signing_input = format!("{}.{}", header_b, payload_b);
    verifying_key
        .verify(signing_input.as_bytes(), &sig)
        .map_err(|_| ProvError("signature invalid".into()))?;

    let payload: Value = serde_json::from_slice(&b64u_decode(payload_b)?)
        .map_err(|e| ProvError(format!("payload parse: {e}")))?;

    validate_payload(&payload)?;

    if header.get("kid").and_then(|v| v.as_str())
        != payload.get("agent_key_id").and_then(|v| v.as_str())
    {
        return err("header.kid != payload.agent_key_id (§3)");
    }

    let id = format!("sha256:{}", sha256_hex(jws.as_bytes()));
    Ok(Receipt {
        jws: jws.to_string(),
        payload,
        id,
    })
}
